wurst/file/FileIO.wurst

```diff
@@ -0,0 +1,105 @@
+package FileIO
+import public ChunkedString
+import ErrorHandling
+import ObjectIdGenerator
+import AbilityObjEditing
+import ObjectIds
+
+/* 	FileIO for Patches 1.31 and up with chunked string support for big data.
+	Reads and writes using preload files by manipulating default ability tooltips.
+
+	Read a file:
+	>	let fileContent = new File("MyFileName.pld").readAsString()
+
+	Write a file:
+	>	new File("MyFileName.pld")..write("MyContent")..close()
+*/
+
+@configurable public constant CHUNKS_PER_FILE = 64
+
+public var CAN_READ_FILES = true
+
+constant DUMMY_ABIL_ID = compiletime(ABIL_ID_GEN.next())
+
+public class File
+	var filename = ""
+
+	construct(string filename)
+		this.filename = filename
+
+	function write(player p, string content)
+		let buffer = new ChunkedString()
+		buffer.append(content)
+		write(p, buffer)
+		destroy buffer
+
+	function write(player p, ChunkedString buffer)
+		for i = 0 to buffer.chunkCount
+			let str = buffer.getChunk(i)
+			if validateInput(str) != null
+				error("FileIO(" + filename + ") ERROR: Invalid character |cffffcc00" + validateInput(str) + "|r")
+
+		writePreload(p, buffer)
+
+	function close()
+		destroy this
+
+	function read(player p) returns ChunkedString
+		let out = new ChunkedString(DEFAULT_CHUNK_SIZE)
+		if localPlayer == p
+			readPreload(out)
+		return out
+
+	function readAsString(player p) returns string
+		let out = new ChunkedString(DEFAULT_CHUNK_SIZE)
+		if localPlayer == p
+			readPreload(out)
+		return out.readChunk()
+
+	private function writePreload(player p, ChunkedString buffer)
+		if localPlayer == p
+			PreloadGenClear()
+			PreloadGenStart()
+
+			for i = 0 to buffer.chunkCount
+				if (i >= CHUNKS_PER_FILE)
+					error("FileIO(" + filename + ") ERROR: String exceeds max length (" + (CHUNKS_PER_FILE * buffer.chunkSize).toString() + ").|r")
+
+				let chunk = buffer.getChunk(i)
+				Preload("\" )\ncall BlzSetAbilityTooltip('" + DUMMY_ABIL_ID.toRawCode() + "', \"" + chunk + "\", " + i.toString() + ")\n//")
```

Comment:
The sense in formatting to hex is to keep the length of the chuck number (the level) and the length of the string constant, since the length of the line is limited. So, when it's constant, the exact max size of the chunk can be determined. An integer should be prefixed with `$` for the Wc3 to read it as a hex number. The hex representation is also shorter for big integers, and anyway we have to consider the maximum possible length of the integer when we set the maximum chunk size.

Reply:
I don't think that's necessary